demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c

```diff
@@ -0,0 +1,1384 @@
+                        eventMsg.eventId = OtaAgentEventStart;
+                        OTA_SignalEvent( &eventMsg );
+                    }
+                }
+            }
+
+            if( mqttSessionEstablished == true )
+            {
+                /* Loop to receive packet from transport interface. */
+                mqttStatus = MQTT_ProcessLoop( &mqttContext, OTA_EXAMPLE_TASK_DELAY_MS );
+
+                /* Get OTA statistics for currently executing job. */
+                OTA_GetStatistics( &otaStatistics );
+
+                if( mqttStatus == MQTTSuccess )
+                {
+                    LogInfo( ( " Received: %u   Queued: %u   Processed: %u   Dropped: %u",
+                               otaStatistics.otaPacketsReceived,
+                               otaStatistics.otaPacketsQueued,
+                               otaStatistics.otaPacketsProcessed,
+                               otaStatistics.otaPacketsDropped ) );
+                }
+                else
+                {
+                    LogError( ( "MQTT_ProcessLoop returned with status = %s.",
+                                MQTT_Strerror( mqttStatus ) ) );
+
+                    /* Disconnect from broker and close connection. */
+                    disconnect();
+
+                    /* Set connection flag to false. */
+                    mqttSessionEstablished = false;
+
+                    /* Suspend OTA operations. */
+                    otaRet = OTA_Suspend();
+
+                    if( otaRet != OtaErrNone )
+                    {
+                        LogError( ( "OTA failed to suspend. "
+                                    "StatusCode=%d.", otaRet ) );
+                    }
+                    else
+                    {
+                        while( ( state = OTA_GetState() ) != OtaAgentStateSuspended )
+                        {
+                            /* Wait for OTA Library state to suspend */
+                            sleep( OTA_EXAMPLE_TASK_DELAY_MS );
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Entry point of demo.
+ *
+ * This example initializes the OTA library to enable OTA updates via the
+ * MQTT broker. It simply connects to the MQTT broker with the users
+ * credentials and spins in an indefinite loop to allow MQTT messages to be
+ * forwarded to the OTA agent for possible processing. The OTA agent does all
+ * of the real work; checking to see if the message topic is one destined for
+ * the OTA agent. If not, it is simply ignored.
+ */
+int main( int argc,
+          char ** argv )
+{
+    ( void ) argc;
+    ( void ) argv;
+
+    /* Return error status. */
+    int returnStatus = EXIT_SUCCESS;
+
+    /* Initialize semaphore for buffer operations. */
+    if( sem_init( &bufferSemaphore, 0, 1 ) != 0 )
```

Comment:
Should there be clean-up operations for the semaphore and thread (running the OTA agent) after the new image is received and activated?

Reply:
https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1418/commits/d7f790141f02f7ca66f7142b099b2897ef70af3c